### ISSUE TITLE
Poor performance and the solutions that emerge

### ISSUE THREAD

**Issue by homonoidian** [homonoidian/ven#2]
Ven is currently approximately 100x (if not more) slower than CPython. That has to change! Crystal is fast. 

Seriously, it's amazingly, beautifully fast -- considering it gives the developer, amongst the other things, a garbage collector and an at-times unnoticeable, elegant type system.

What I'm saying is this: it's my choices that were incorrect. I am the one who one should blame for the performance issues. Try to give `examples/lexer.ven` a reasonably large expression.

Ven is a prototype, though. And, as I am learning to develop things, a curious interpretation infrastructure emerges.

The node visitor will still be there. It will be used partly for *read macros* (the first curiosity), partly for *macros* (the second), and partly for *comptimes* and *comptime macros* (shoutout to Zig, I guess). Their bodies (whatever this means) shall look and work just if they were running at runtime (including side effects), but running at compile-time instead. Macros & Co. as powerful as Lisp's. Everyone's dream!

That said, the runtime will be a (hopefully) fast virtual machine.

These ideas are a little convoluted. One of the many examples of the speed up they may provide, though, is what follows.

At read-macro-time, a regex transforms into a quote; at macro-time, this quote compiles into a bunch of quotes that implement the finite state machine this regex represents. All this so that at runtime, only the specialized state machine runs. No regex compilation overhead.

I've seen this somewhere (in D, it seems). And it'll be cool to have, considering Ven's an interpreted language. You know what I mean.

As I've said, these ideas are a little convoluted. And I feel I'm not ready yet to implement them **all**. I don't know many things. But I'll try, step by step, one feature at a time.

You can, too, though I don't know for sure how to manage a project with git.

**Comment by AmitPress** [comment]
Hi, can I ask you something? Is this a hobby project? or you are seriously wanting to solve a/some problem with your new language?

**Comment by homonoidian** [comment]
Mostly a hobby project, yes. I'm testing out some ideas and learning the craft. There isn't any concrete problem I'm trying to solve with/through Ven. But there is, I think, a general one: safely handling things that are considered dangerous and/or experimental in other languages (e.g., read macros, compile-time evaluation, etc.)

**Comment by AmitPress** [comment]
Yeah, I kind of sensed that. I was just watching your project because I am willing to develop a compiler for a specific purpose. Anyway, best of luck for your project. 😄

**Comment by homonoidian** [comment]
Thanks. Best of luck for your project too :)

**Comment by homonoidian** [comment]
This was more of an abstract issue, so me closing it a) doesn't mean Ven became faster, and b) doesn't mean the solutions I've proposed here were merged into the language successfully. 

Today, Ven is interpreted by a slowish (again, mostly because of my sloppiness) bytecode machinery. Even now it's not Python fast; but it's not extremely slow too (being about 20x slower than Python, although this is just a rough estimate). That'll do. I mean, I have better things to do than trying to squeeze out every little droplet of performance from Ven.

As for the regex thing, again, there are better things to do _right now_.

So this issue is more or less closed, yeah. :slightly_smiling_face:
